sui_programmability/examples/staking/sources/stake_utils.move

```diff
@@ -0,0 +1,95 @@
+module staking::stake_utils {
+    use sui::sui_system::{Self, SuiSystemState};
+    use sui::validator;
+    use sui::validator_set;
+    use std::vector;
+
+    /// The given validator is not part of the validator set
+    const EValidatorNotFound: u64 = 0;
+
+    /// The given rank is larger than the validator set
+    const ERankOutOfBounds: u64 = 1;
+
+    /// Total stake info and stake ranking, if the epoch ended right now
+    struct ValidatorInfo has store, drop, copy {
+        /// Address of the validator
+        addr: address,
+        /// Current total stake of the validator
+        stake: u64,
+        /// 0-indexed rank of the validator
+        rank: u64,
+    }
+
+    /// Return the current stake info and stake rank for `validator`.
+    /// This represents what the validator's stake and rank would be if the current epoch ended right now
+    public fun next_epoch_info(validator: address, state: &SuiSystemState): (address, u64, u64) {
+        let stakes = next_epoch_stakes(state);
+        let info = find_validator_info(validator, &stakes);
+        (info.addr, info.stake, info.rank)
+    }
+
+    /// Return the current stake info and stake rank for the validator with the given `rank`
+    /// This represents what the validator's stake and rank would be if the current epoch ended right now
+    public fun next_epoch_info_for_rank(rank: u64, state: &SuiSystemState): (address, u64, u64) {
+        let stakes = next_epoch_stakes(state);
+        assert!(rank < vector::length(&stakes), ERankOutOfBounds);
+        let info = vector::borrow(&stakes, rank);
+        (info.addr, info.stake, info.rank)
+    }
+
+    /// Return live stake info based on pending delegations and withdrawals
+    /// Intended for client usage, helpful if you want to see how your assigned
+    /// validator is doing in the middle of an epoch
+    /// Note: assumes static validator set
+    public fun next_epoch_stakes(state: &SuiSystemState): vector<ValidatorInfo> {
+        let validators = validator_set::active_validators(sui_system::validators(state));
+        let next_epoch_stakes = vector[];
+        let i = 0;
+        let num_validators = vector::length(validators);
+        while (i < num_validators) {
+            let validator = vector::borrow(validators, i);
+            let addr = validator::sui_address(validator);
+            let stake = validator::total_stake(validator) + validator::pending_stake_amount(validator) - validator::pending_withdraw(validator);
```

Comment:
pending_stake_amount and pending_withdraw only give you numbers for the pending self stakes, not delegated stake. And unfortunately we don't have getters for pending delegated stakes.